ops/model.py

```diff
@@ -1324,7 +1368,227 @@ def list_files(self, path: str, *, pattern: str = None,
+            # Destination results: /dst/foobar.txt
+
+            # copy a directory
+            container.pull_path('/foo', '/dst')
+            # Destination results: /dst/foo/bar/baz.txt, /dst/foo/foobar.txt
+
+            # copy a directory's contents
+            container.pull_path('/foo/*', '/dst')
+            # Destination results: /dst/bar/baz.txt, /dst/foobar.txt
+
+            # copy multiple files
+            container.pull_path(['/foo/bar/baz.txt', 'quux.txt'], '/dst')
+            # Destination results: /dst/baz.txt, /dst/quux.txt
+
+            # copy a file and a directory
+            container.pull_path(['/foo/bar', '/quux.txt'], '/dst')
+            # Destination results: /dst/bar/baz.txt, /dst/quux.txt
+
+        Args:
+            source_path: A single path or list of paths to pull from the remote system.  The
+                paths can be either a file or a directory but must be absolute paths.  If
+                source_path is a directory, the directory base name is attached to the destination
+                directory - i.e.  the source path directory is placed inside the destination
+                directory.  If a source path ends with a trailing "/*" it will have its *contents*
+                placed inside the destination directory.
+            dest_dir: Local destination directory inside which the source dir/files will be
+                placed.
+        """
+        if os.name == 'nt':
+            raise RuntimeError('Container.pull_path is not supported on Windows-based systems')
+
+        source_paths = source_path if isinstance(source_path, list) else [source_path]
+        source_paths = [str(p) for p in source_paths]
+        dest_dir = str(dest_dir)
+
+        errors = []
+        for source_path in source_paths:
+            try:
+                for f in Container._list_recursive(self.list_files, source_path):
+                    srcpath, dstpath = f.path, self._build_destpath(f.path, source_path, dest_dir)
+                    os.makedirs(os.path.dirname(dstpath), exist_ok=True)
+                    with self.pull(srcpath, encoding=None) as src:
+                        with open(dstpath, 'wb') as dst:
+                            shutil.copyfileobj(src, dst)
+            except Exception as err:
+                errors.append((source_path, err))
+        if len(errors) > 0:
+            raise MultiPushPullError('failed to pull one or more files', errors)
+
+    @staticmethod
+    def _build_fileinfo(path: str) -> ops.pebble.FileInfo:
+        """Constructs a FileInfo object by stat'ing a local path."""
+        if os.path.islink(path):
+            ftype = pebble.FileType.SYMLINK
+        elif os.path.isdir(path):
+            ftype = pebble.FileType.DIRECTORY
+        elif os.path.isfile(path):
+            ftype = pebble.FileType.FILE
+        else:
+            ftype = pebble.FileType.UNKNOWN
+
+        import grp
+        import pwd
```

Comment:
Is there a particular reason we're importing this late?

Reply:
They don't exist on non-unixy systems.

Reply:
Fair 'nuf :)